### ISSUE TITLE
k8s custom resource definition for repository locations

### ISSUE THREAD

**Issue by peay** [dagster-io/dagster#7139]
## Use Case

When deploying Dagster on kubernetes, one can configure repository locations through the `workspace.yaml` file. In the official Helm chart, it can be configured by listing in the value definitions the repository locations.

This requires updating the Dagster deployment to add new repository locations. Something more idiomatic for kubernetes deployment would be to allow users to define `DagsterWorkspaceRepositoryLocation` custom kubernetes resources, and have the Dagster components pick those up.

A good use case for this is 'branch deployments', where you have a repository with pipeline code, and would like to be able to run each branch in a shared Dagster deployment. A CI pipeline could create a `DagsterWorkspaceRepositoryLocation` custom resource for the branch to make it available in the Dagster instance.

## Ideas of Implementation

The `DagsterWorkspaceRepositoryLocation` custom resource definition could mimic the existing schema for a repository location in `workspace.yaml`.

We would need a way to configure Dagster to look for these custom resources. Incorporating this into `workspace.yaml` might be the easiest way. There would be several options:

- As a regular `load_from` entry:

  ```yaml
  load_from:
  - k8s_custom_resources:
       namespace: ...   # Namespace where to look for custom resources
       selector:  # Optional label-based selector on which custom resources to consider
          dagster-deployment: dagster-dev
  ```

  Although this might be natural, it is a bit inconsisent in that this single item might yield multiple repository locations if there are multiple custom resources. Currently, each item in `load_from` maps to a single repository location only.

- As a new top-level block - e.g., `load_repository_locations_from:` - where each item can yield multiple repository locations instead of only one. I am not sure if it is worth the additional complexity on the YAML schema, though.

A naive implementation would implement `k8s_custom_resources` using the k8s Python client to list custom resources in the prescribed namespace, possibly with the selectors. This would run whenever the workspace is refreshed. Separate components (Dagster daemon, Dagit) would each list resources independently. If there are a lot of custom resources, it might end up being a bit slow and/or generate load on the kubernetes API server. However, it is likely to be enough for most use cases already.

A more refined implementation would have a small, dedicated component that acts as a controller for the custom resource, and can expose the repository locations via a gRPC API as a user code deployment can do today for a single repository location. The controller can subscribe to kubernetes CRUD events for the custom resource and update its internal state accordingly to prevent all components from having to keep listing resources periodically.

## Additional Info

---
#### Message from the maintainers:
Excited about this feature? Give it a :thumbsup:. We factor engagement into prioritization.

**Comment by johannkm** [comment]
Thanks for the detailed proposal! If I'm understanding correctly, the main benefits of this would be:

- repo locations could be added and removed without a `helm upgrade` on Dagit/Daemon
- repo locations could be added and removed individually without having a centralized list, since in the branch approach no one branch would have a list of all the repo locations

No ETA yet, but we're working on bringing in the ability to add and remove repo locations easily. Here's a preview of how it works in Dagster Cloud: https://docs.dagster.cloud/guides/adding-code. I'm curious if this would solve most of your use case, or if there would be other benefits to you of having CRDs.

**Comment by peay** [comment]
Yes, that's correct. Using an API to configure origins would be quite good as well as long as that API allows CRUD at the level of individual origins.

CRD might fit slightly better in the kubernetes workflow: a branch deployment can simply deploy the gRPC server with the code for the branch along with the CRD to define the corresponding origin.

With the API, such a deployment would have to resort to, for instance, a post-upgrade or post-delete Helm hook that fires up a Dagster API client to add/update/delete the origin via the API, which would be more cumbersome and less "batteriers-included". If for some reason a post-delete hook failed to run when the corresponding Helm release was undeployed, then we might leave origins hanging in Dagster, etc.

**Comment by peay** [comment]
However, API and CRD may not be incompatible: if Dagster provides a clean API to add/remove origins, then it should be easy to implement a CRD controller outside of Dagster which can be used if needed and uses the API under the hood.

Today, it is a bit hard to implement such a controller separately because Dagit/Daemon do not reload the `workspace.yaml` after startup, so even if `workspace.yaml` has been modified, these two pods need to be restarted to take into account any modifications.

**Comment by peay** [comment]
> No ETA yet, but we're working on bringing in the ability to add and remove repo locations easily. Here's a preview of how it works in Dagster Cloud: https://docs.dagster.cloud/guides/adding-code. I'm curious if this would solve most of your use case, or if there would be other benefits to you of having CRDs.

@johannkm now that Dagster Cloud is out, do you have any plans to back port this API to add/remove repo locations? It'd be very useful to implement a robust kubernetes custom resource controller for managing workspace (much more so than configmaps).

**Comment by peay** [comment]
@johannkm @rexledesma I wanted to touch base on this, as I was wondering whether to open-source my custom resource controller for Dagster deployments -- however, using the YAML file with a configmap is still a bit unreliable due to kubernetes delays with updating configmaps. Are there any plans to backport these APIs to manage repo locations?

**Comment by johannkm** [comment]
Sorry I missed that prior message. Bringing the add/remove repo location api to open source would be part of a larger effort- it means OSS Dagster has to be able to spin up grpc servers for you (where currently we rely on the helm chart/etc.). We don't plan to do this soon unfortunately.

If you're up for open sourcing your CRDs I think that'd be great! There's certainly users that we could point towards it.

**Comment by peay** [comment]
Thanks @johannkm -- I actually had in mind something with less coupling, let me know if that would still make sense for OSS as I think that would be easier to implement.

The idea would be as follows:

- Have APIs in Dagster to register/un-register locations without having to go through `workspace.yaml` (e.g., storing those in the database). This does not include any deployment, but only to register locations given the URLs of gRPC servers.

- Have users bring their own gRPC servers (e.g., using the Helm chart), and a `DagsterDeployment` custom resource alongside. This is the most reasonable option for OSS anyway as there's so much variety in how folks would want to handle that.

- CRD controller can register/un-register the corresponding location using the APIs instead of hacks with `workspace.yaml` mounted through a configmap.

**Comment by peay** [comment]
@johannkm any thoughts on this model? Do you see a path to adding APIs in Dagster OSS to register/un-register locations that would allow to accommodate this? Thanks!
